functions taking calldata args should be assignable to function pointers of same type (#13227)
ethereum/solidity: docs/control-structures.rst

    external functions happen after any changes to state variables in your contract
    so your contract is not vulnerable to a reentrancy exploit.

> .. warning::

Reviewer: I think a note is enough. It's a detail that's good to have some understanding of but you don't have to be on guard and actively watch for it.
```suggestion
.. note::
```

> .. warning::

Reviewer: Wouldn't this fit better under [Types > Function Types](https://docs.soliditylang.org/en/latest/types.html#function-types)? It even has a paragraph about assigning function types.

Author: Changed the location of this `note` in the latest push.

>     As stated previously, for an external call, all arguments have to be copied to memory.
>     As a consequence of that, if an external function has ``calldata`` parameters,
>     they will be internally converted by the compiler to ``memory`` parameters.
>     In practice, there is no actual difference between these two data locations for arguments of an external call.
>     Thus, external function pointers cannot be assigned, passed as arguments or used in ``abi.encodeCall( )``.
>     Attempts of these operations will result in type related errors.
>     Hence, in such cases, it is recommended to use external functions with ``memory`` parameters to circumvent the issue.

Reviewer: I think that the explanation of why it's this way could be a bit clearer. I'd also put what happens first and only then explain why. How about this?

```suggestion
    External functions with ``calldata`` parameters are incompatible with external function types with ``calldata`` parameters.
    They are compatible with the corresponding types with ``memory`` parameters instead.
    For example, there is no function that can be pointed at by a value of type ``function (string calldata) external`` while
    ``function (string memory) external`` can point at both ``function f(string memory) external {}`` and
    ``function g(string calldata) external {}``.

    This is because for both locations the arguments are passed to the function in the same way.
    The caller cannot pass its calldata directly to an external function and always ABI-encodes the arguments into memory.
    Marking the parameters as ``calldata`` only affects the implementation of the external function and is
    meaningless in a function pointer on the caller's side.
```

Reviewer: What about this one? It was marked as resolved but the text is still pretty much as it was before. I think it's a bit too vague about what's actually happening.

Author: Made relevant changes in the latest push.

    external functions happen after any changes to state variables in your contract
    so your contract is not vulnerable to a reentrancy exploit.

> .. warning::

Reviewer 2: ```suggestion
.. warning::
 As stated previously, for an external call, all arguments have to be copied to memory. As a consequence of that, if an external function has ``calldata`` parameters, they will be internally (compulsorily ?) converted by the compiler to ``memory`` parameters. In practice, there is no actual difference between these two data locations for arguments of an external call. Thus, external function pointers cannot be assigned, passed as arguments or used in ``abi.encodeCall( )``. Attempts of these operations will result in type related errors. This can be seen in function  ``main()`` in the following example.

.. code-block:: solidity
// SPDX-License-Identifier: GPL-3.0
        pragma solidity >=0.6.2 <0.9.0;

        contract C {
            function f(function (string calldata) external) external {}

            function g(string calldata) external {}

            function main() external {
                function (string calldata) external ptr = this.g;
                abi.encodeCall(this.f, (this.g));
                this.f(this.g);
            }

        }

Although function ``g`` is declared to have ``calldata`` parameters, the function pointer ``this.g`` has its type converted to ``function (string memory) external`` and then the above contract will produce the following errors when compiled:

Error: Type function (string memory) external is not implicitly convertible to expected type function (string calldata) external.
 --> test.sol:6:9:
  |
6 |         function (string calldata) external ptr = this.g;
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Error: Cannot implicitly convert component at position 0 from "function" to "function".
 --> test.sol:7:32:
  |
7 |         abi.encodeCall(this.f, (this.g));
  |                                ^^^^^^^^

Error: Invalid type for argument in function call. Invalid implicit conversion from function (string memory) external to function (string calldata) external requested.
 --> test.sol:8:16:
  |
8 |         this.f(this.g);
  | 
                 ^^^^^^
In this case, it is recommended to use external functions with ``memory``parameters to circumvent the issue.  
```

Reviewer: @matheusaaguiar I think that the error messages make this a bit too verbose. Overall, this is just a corner case that's not very useful. Users should be made aware of it but it does not have to be prominent.

Maybe it does not even need an example. At least this test case is not a very good one. For example it uses `abi.encodeCall()` which is completely irrelevant and only makes it more complex. If I were to include an example, it would probably be something closer to this:
```solidity
// SPDX-License-Identifier: GPL-3.0
pragma solidity >=0.6.2 <0.9.0;

contract C {
    function calldataFunction(string calldata) external {}
    function memoryFunction(string memory) external {}

    function assignPtr() external view {
        function (string calldata) external calldataPtr;
        function (string memory) external memoryPtr;

        // OK
        memoryPtr = this.calldataFunction;
        memoryPtr = this.memoryFunction;

        // Error: Type function (string memory) external is not implicitly
        // convertible to expected type function (string calldata) external.
        //calldataPtr = this.calldataFunction;
        //calldataPtr = this.memoryFunction;
    }
}
```

Reviewer 2: I thought that it was getting too long to be just a warning, but went ahead anyway because I find it easier to remove and trim than to add more text :)
Let's just remove the example and the error messages then.

    External function pointers only take memory arguments, even if the function signature uses calldata.
    This is because for external function pointers, you are only interested in the calling side and not in the called side.

>     For the caller, it does not make sense to even distinguish `function (string calldata) external` and `function (string memory) external`.

Reviewer: It's RST, not Markdown. You need double backticks:
```suggestion
    For the caller, it does not make sense to even distinguish ``function (string calldata) external`` and ``function (string memory) external``.
```

Reviewer 2: Ok.

    So, effectively, for external function pointers taking calldata arguments, the compiler treats them as if they had memory arguments instead.
    So, if you do supply calldata arguments to external function pointers,

>     you cannot assign them, pass into a function of use in ``abi.encodeCall()`` even when the target type exactly matches the

Reviewer: ```suggestion
    you cannot assign them, pass into a function or use in ``abi.encodeCall()`` even when the target type exactly matches the
```

    type of the said function pointer.

>     For eg. All 3 lines of code in the following ``main()`` function will result in errors.

Reviewer: ```suggestion
    For example, all 3 lines of code in the following ``main()`` function will result in errors.
```

    external functions happen after any changes to state variables in your contract
    so your contract is not vulnerable to a reentrancy exploit.

.. warning::
    The type of this.g takes a memory argument, even if the function signature uses calldata.
    This is because for external function pointers, you are only interested in the calling side and not in the called side.
    For the caller, it does not make sense to even distinguish

>     function (string calldata) external and function (string memory) external.

Reviewer 2: The text here can be rephrased to present a more informed situation to the reader. The lines here are [comments](https://github.com/ethereum/solidity/issues/12778#issuecomment-1169917941) from an ongoing discussion. 
A better approach maybe would be to first explain the error (the `calldata` arguments of external function pointers get converted internally to `memory` arguments) then explain the reasons (interest in calling side not called side...). After that, suggestion of how to amend the situation and examples of the problem can be showed to conclude.

>     The type of this.g takes a memory argument, even if the function signature uses calldata.
>     This is because for external function pointers, you are only interested in the calling side and not in the called side.
>     For the caller, it does not make sense to even distinguish
>     function (string calldata) external and function (string memory) external.
> 
>     So, effectively, for external functions taking calldata arguments, the compiler treats them as if they had memory arguments instead.
> 
>     You cannot assign them, pass into a function of use in abi.encodeCall() even when the target type exactly matches the
>     type of the function pointer.
> 
>     For eg. All 3 lines of code in the following main() function will result in errors.

Reviewer: ```suggestion
    The type of ``this.g`` takes a memory argument, even if the function signature uses calldata.
    This is because for external function pointers, you are only interested in the calling side and not in the called side.
    For the caller, it does not make sense to even distinguish
    function (string calldata) external and function (string memory) external.

    So, effectively, for external functions taking calldata arguments, the compiler treats them as if they had memory arguments instead.

    You cannot assign them, pass into a function of use in ``abi.encodeCall()`` even when the target type exactly matches the
    type of the function pointer.

    All 3 lines of code in the following ``main()`` function will result in errors.
```

    .. code-block:: solidity

        // SPDX-License-Identifier: GPL-3.0
        pragma solidity >=0.6.2 <0.9.0;

        contract C {
            function f(function (string calldata) external) external {}

            function g(string calldata) external {}

            function main() external {
                function (string calldata) external ptr = this.g;
                abi.encodeCall(this.f, (this.g));
                this.f(this.g);
            }

        }

>     Here, we must observe that function (string calldata) external is a type that is not really useful.

Reviewer: ```suggestion
    Here, we must observe that ``function (string calldata) external`` is a type that is not really useful.
```
